x/upgrade/abci_test.go

```diff
@@ -185,6 +209,162 @@ func (s *TestSuite) TestPlanStringer() {
   Info: `, Plan{Name: "test", Height: 100}.String())
 }
 
+func (s *TestSuite) VerifyNotDone(newCtx sdk.Context, name string) {
+	s.T().Log("Verify that upgrade was not done")
+	height := s.keeper.GetDoneHeight(newCtx, name)
+	s.Require().Zero(height)
+}
+
+func (s *TestSuite) VerifyDone(newCtx sdk.Context, name string) {
+	s.T().Log("Verify that the upgrade plan has been executed")
+	height := s.keeper.GetDoneHeight(newCtx, name)
+	s.Require().NotZero(height)
+}
+
+func (s *TestSuite) VerifySet() {
+	s.T().Log("Verify if the skip upgrade has been set")
+	s.Require().NotNil(viper.GetIntSlice(s.FlagUnsafeSkipUpgrade))
+}
+
+func (s *TestSuite) VerifyConversion(skipUpgrade []int) {
+	skipUpgradeHeights := s.keeper.ConvertIntArrayToInt64(skipUpgrade)
+	s.Require().Equal(reflect.TypeOf(skipUpgradeHeights).Elem().Kind(), reflect.Int64)
+}
+
+func (s *TestSuite) TestContains() {
+	skipUpgradeHeights := s.keeper.ConvertIntArrayToInt64(viper.GetIntSlice(s.FlagUnsafeSkipUpgrade))
+	s.T().Log("case where array contains the element")
+	present := s.keeper.Contains(skipUpgradeHeights, -1)
+	s.Require().True(present)
+
+	s.T().Log("case where array doesn't contain the element")
+	present = s.keeper.Contains(skipUpgradeHeights, 4)
+	s.Require().False(present)
+}
+
+func (s *TestSuite) TestSkipUpgradeSkippingBoth() {
+	newCtx := sdk.NewContext(s.cms, abci.Header{Height: s.ctx.BlockHeight() + 1, Time: time.Now()}, false, log.NewNopLogger())
+	req := abci.RequestBeginBlock{Header: newCtx.BlockHeader()}
+	err := s.handler(s.ctx, SoftwareUpgradeProposal{Title: "prop", Plan: Plan{Name: "test", Height: s.ctx.BlockHeight() + 1}})
+	s.Require().Nil(err)
+
+	var i int64
+
+	s.T().Log("Verify if skip upgrade flag clears upgrade plan in both cases")
+	viper.Set(s.FlagUnsafeSkipUpgrade, []int{int(s.ctx.BlockHeight() + 1), int(s.ctx.BlockHeight() + 10)})
+	s.VerifySet()
+
+	s.VerifyConversion(viper.GetIntSlice(s.FlagUnsafeSkipUpgrade))
+
+	for i = 1; i <= s.ctx.BlockHeight(); i++ {
```

Comment:
It would be nice to make it more clear we tested the first upgrade here... I was confused why you stopped one block early, but then saw `newCtx.WithBlockHeight(s.ctx.BlockHeight() + i)`. which actually doesn't make too much sense... (doing from blockheight+1 to double blockheight).

I would:
* rename `i` to `h`
* make `h` exactly the block height you want `newCtx.WithBlockHeight(h)`
* adjust the for loop to whatever blocks you want to cover.

Not that this code is wrong persay, just not clear.
(Maybe you do want to test if the flag is set for a block that has no upgrade, that is a fine test, and you do it here, probably, it is just not clear)

Reply:
Over second thought,I feel this looping can be avoided and BeginBlock can be run only for the heights mentioned in the proposals. Do you think I'll miss some case if I do that, I don't see any issues

Reply:
It seems fine.
I would make sure to run every case in the following matrix at least once:

* upgrade=no, skip=no
* upgrade=yes, skip=no
* upgrade=no, skip=yes
* upgrade=yes, skip=yes

Beyond that, no need to check any blocks (and this matrix doesn't have to be run in every text case, but each at least once in the suite).